pipeline/train/train.sh

```diff
@@ -48,7 +129,37 @@ echo "### Training ${model_dir}"
   --keep-best \
   --log "${model_dir}/train.log" \
   --valid-log "${model_dir}/valid.log" \
-  "${extra_params[@]}"
+  "${extra_params[@]}" &
+
+# keep track of the process ID of the child
+marian_pid=$!
+
+wait $marian_pid
+wait $marian_pid # Yes, twice. See explanation below.
+marian_exit_status=$?
+# Wait stops waiting under two conditions.
+# 1. The process being waited for finishes. In that case, $? is its exit status.
+# 2. A signal handler returns. In that case, $? is 128 plus the signal number.
+# A second wait sets does no harm in the first case and sets $? correctly in the
+# second.
 
-echo "### Model training is completed: ${model_dir}"
-echo "###### Done: Training a model"
+# Now we interpret the exit status.
+if [ "$marian_exit_status" == 143 ]; then # 143 = 128 + 15 => SIGTERM
+    echo "### Model training was interrupted: ${model_dir}"
+    echo "###### Not done: Training a model"
+    # Do not cancel the job array! We want to continue in the next slot
+else
+    if [ "$marian_exit_status" == 0 ]; then
+        echo "### Model training is completed: ${model_dir}"
+        echo "###### Done: Training a model"
+    else
+        echo "### Model training failed: ${model_dir}"
+        echo "###### Not done: Training a model"
+    fi;
+    if [[ "${SLURM_JOBID:-''}" != "" ]] ; then
```

Comment:
this script is not supposed to know that it's run on slurm. It can be run locally or on any other cluster supported by snakemake.

Reply:
Checking whether or not this is a slurm script is excactly what this line of code does. If it's not a slurm script, why would SLURM_JOBID be anything other than either not set or empty? Why on earth would any non-slurm environment have SLURM_JOBID set?

Reply:
Let me try to explain my point better: this is a valid code and it definitely checks things correctly, but this script should not cancel any jobs if we look at it from a current architecture perspective. This script should not be aware of where it is being run. Job scheduling, status checking and cancelling should happen in the profile. 

However, if we'll try to implement training using job arrays, this hack might be handy. I tried to handle this cancelling in the [status script](https://github.com/mozilla/firefox-translations-training/blob/main/profiles/slurm-csd3/status.py) when the job array step is completed, but it's too late and the next step had already started. After that snakemake deletes output of the completed step and breaks everything. This is a bit complicated. But again we'll need this only if we decide to hack job arrays or run this script separately from snakemake.